fdbclient/ManagementAPI.actor.cpp

```diff
@@ -465,6 +467,177 @@ bool isCompleteConfiguration(std::map<std::string, std::string> const& options)
+                                        std::map<std::string, std::string> newConfig,
+                                        bool creating) {
+	if (creating) {
+		if (newConfig.count(encryptionAtRestModeConfKey.toString()) != 0) {
+			EncryptionAtRestMode encryptMode = EncryptionAtRestMode::fromValueRef(
+			    ValueRef(newConfig.find(encryptionAtRestModeConfKey.toString())->second));
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED ||
+			    encryptMode.mode == EncryptionAtRestMode::CLUSTER_AWARE) {
+				// In these cases the tenantMode does not matter
+				return true;
+			}
+
+			// check if the tenant mode is being set during configure new (otherwise assume tenants are disabled)
+			TenantMode tenantMode = TenantMode::DISABLED;
+			if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+				tenantMode = TenantMode::fromValue(ValueRef(newConfig.find(tenantModeConfKey.toString())->second));
+			}
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("TenantMode", tenantMode.toString());
+			if (tenantMode != TenantMode::REQUIRED) {
+				// For domain aware encryption only the required tenant mode is currently supported
+				TraceEvent(SevWarnAlways, "InvalidEncryptAndTenantConfiguration")
+				    .detail("EncryptMode", encryptMode.toString())
+				    .detail("TenantMode", tenantMode.toString());
+				return false;
+			}
+			return true;
+		}
+	} else {
+		if (newConfig.count(tenantModeConfKey.toString()) != 0) {
+			ASSERT(oldConfiguration.present());
+			// If the tenant mode is being changed then make sure it obeys the current cluster encryption state
+			EncryptionAtRestMode encryptMode = oldConfiguration.get().encryptionAtRestMode;
+
+			if (encryptMode.mode == EncryptionAtRestMode::DISABLED) {
+				// In this case the tenantMode does not matter
+				return true;
+			}
+			TenantMode oldTenantMode = oldConfiguration.get().tenantMode;
+			TenantMode newTenantMode = TenantMode::fromValue(newConfig.find(tenantModeConfKey.toString())->second);
+			TraceEvent(SevDebug, "EncryptAndTenantModes")
+			    .detail("EncryptMode", encryptMode.toString())
+			    .detail("OldTenantMode", oldTenantMode.toString())
+			    .detail("NewTenantMode", newTenantMode.toString());
+
+			if (encryptMode.mode == EncryptionAtRestMode::DOMAIN_AWARE) {
+				// For domain aware encryption only the required tenant mode is currently supported (changing to any
+				// other mode is disallowed)
+				ASSERT(oldTenantMode == TenantMode::REQUIRED);
+				if (newTenantMode != TenantMode::REQUIRED) {
+					TraceEvent(SevWarnAlways, "InvalidEncryptAndTenantConfiguration")
+					    .detail("EncryptMode", encryptMode.toString())
+					    .detail("OldTenantMode", oldTenantMode.toString())
+					    .detail("NewTenantMode", newTenantMode.toString());
+					return false;
+				}
+			} else if (oldTenantMode != TenantMode::REQUIRED && newTenantMode == TenantMode::REQUIRED) {
```

Comment:
My suggestion would be to not do any tenant mode specific checks here. Changes from optional/disable -> required should be handled in a separate block because they don't relate to encryption at all, and those are currently part of the tenant implementation plans.

Then I think this function can be written more simply:

Get the encryption mode (from the config params or the existing config). If that mode is domain aware, then get the tenant mode (from config params or the existing config). If tenant mode is not required, fail. Otherwise, you can succeed.

I'm not sure if it's being handled elsewhere currently, but if it isn't you could also check for the case where encryption mode is being set outside of configure new.

Reply:
done